### ISSUE TITLE
HealthCheckService execution leads to a deadlock in the DI stack

### ISSUE THREAD

**Issue by TheAzureGuy** [dotnet/runtime#70412]

**Comment by rafikiassumani-msft** [comment]
We are assuming that this was fixed by https://github.com/dotnet/runtime/issues/35986. Can you assess if there is a need to backport the fix?

**Comment by halter73** [comment]
> Other blocked threads exhibit this stack:

Are there any other threads that have a similar stack? It looks like there's probably a deadlock in some service's constructor like there was with the `IHttpClientFactory` issue linked above. A per-singleton-lock like the one introduced by https://github.com/dotnet/runtime/pull/46157 in .NET 6 might fix the issue and possibly should be backported, but fixing the service constructor might also be a workaround.

**Comment by davidfowl** [comment]
We can't back port that fix, it's too big of a change to backport.

> but fixing the service constructor might also be a workaround.

That sounds good if that works. I haven't looked into this issue specifically.

**Comment by TheAzureGuy** [comment]
> > Other blocked threads exhibit this stack:
> 
> Are there any other threads that have a similar stack? It looks like there's probably a deadlock in some service's constructor like there was with the `IHttpClientFactory` issue linked above. A per-singleton-lock like the one introduced by #46157 in .NET 6 might fix the issue and possibly should be backported, but fixing the service constructor might also be a workaround.

Yes, there were many threads (77 out of approx 100) that were blocked with the same stack trace.

**Comment by TheAzureGuy** [comment]
> We can't back port that fix, it's too big of a change to backport.
> 
> > but fixing the service constructor might also be a workaround.
> 
> That sounds good if that works. I haven't looked into this issue specifically.

No worries, we will upgrade to .NET 6 and give it a try.

**Comment by eerhardt** [comment]
@TheAzureGuy - can you let us know if this repros on .NET 6? If it doesn't reproduce on the latest release we can close the issue.
